sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/UnsafeArrayData.java

```diff
@@ -230,7 +233,10 @@ public UTF8String getUTF8String(int ordinal) {
     final long offsetAndSize = getLong(ordinal);
     final int offset = (int) (offsetAndSize >> 32);
     final int size = (int) offsetAndSize;
-    return UTF8String.fromAddress(baseObject, baseOffset + offset, size);
+    MemoryBlock mb = (baseObject instanceof byte[]) ?
+      new ByteArrayMemoryBlock((byte[]) baseObject, baseOffset + offset, size) :
+      new LongArrayMemoryBlock((long[]) baseObject, baseOffset + offset, size);
```

Comment:
shouldn't this logic belong to `UTF8String.fromAddress`?

Reply:
I refactor this and move this logic into `MemoryBlock`.